File: sdk-android/TuneMarketingConsoleSDK/src/main/java/com/tune/ma/experiments/TuneExperimentManager.java
package com.tune.ma.experiments;

import com.tune.ma.eventbus.TuneEventBus;
import com.tune.ma.eventbus.event.TunePlaylistManagerCurrentPlaylistChanged;
import com.tune.ma.eventbus.event.TuneSessionVariableToSet;
import com.tune.ma.experiments.model.TuneExperimentDetails;
import com.tune.ma.experiments.model.TuneInAppMessageExperimentDetails;
import com.tune.ma.experiments.model.TunePowerHookExperimentDetails;
import com.tune.ma.playlist.model.TunePlaylist;
import com.tune.ma.powerhooks.model.TunePowerHookValue;
import com.tune.ma.utils.TuneJsonUtils;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

/**
 * Created by gowie on 1/26/16.
 */
public class TuneExperimentManager {

    public static final String TUNE_ACTIVE_VARIATION_ID = "TUNE_ACTIVE_VARIATION_ID";

    private Map<String, TunePowerHookExperimentDetails> phookExperimentDetails;
    private Map<String, TuneInAppMessageExperimentDetails> inAppExperimentDetails;
    private Set<String> addedActiveVariations;

    public TuneExperimentManager() {
        phookExperimentDetails = new HashMap<String, TunePowerHookExperimentDetails>();
        inAppExperimentDetails = new HashMap<String, TuneInAppMessageExperimentDetails>();
        addedActiveVariations = new HashSet<String>();
    }

    public void onEvent(TunePlaylistManagerCurrentPlaylistChanged event) {
        TunePlaylist activePlaylist = event.getNewPlaylist();

        Map<String, TunePowerHookExperimentDetails> phookExperimentDetailsTemp = new HashMap<String, TunePowerHookExperimentDetails>();
        Map<String, TuneInAppMessageExperimentDetails> inAppMessageExperimentDetailsTemp = new HashMap<String, TuneInAppMessageExperimentDetails>();

        JSONObject experimentDetails = activePlaylist.getExperimentDetails();
        if (experimentDetails == null) {
            return;
        }

        Iterator<String> experimentDetailsIter = experimentDetails.keys();
        String experimentId;
        while (experimentDetailsIter.hasNext()) {
            experimentId = experimentDetailsIter.next();
            JSONObject experiment = TuneJsonUtils.getJSONObject(experimentDetails, experimentId);
            String type = TuneJsonUtils.getString(experiment, TuneExperimentDetails.DETAIL_EXPERIMENT_TYPE_KEY);

            if (type != null && type.equals(TuneExperimentDetails.DETAIL_TYPE_POWER_HOOK)) {
                JSONObject hooks = activePlaylist.getPowerHooks();
                Iterator<String> hooksIter = hooks.keys();
                String hookId;
                while (hooksIter.hasNext()) {
                    hookId = hooksIter.next();
                    // Create the PowerHookValue temporarily so we can get some information from it.
                    TunePowerHookValue hookValue = new TunePowerHookValue();
                    hookValue.mergeWithPlaylistJson(TuneJsonUtils.getJSONObject(hooks, hookId));
                    if (experimentId.equals(hookValue.getExperimentId())){
                        TunePowerHookExperimentDetails details = new TunePowerHookExperimentDetails(experiment, hookValue);
                        phookExperimentDetailsTemp.put(hookId, details);

                        if (addedActiveVariations.contains(details.getCurrentVariantId())) {
                            break;
                        } else {
                            addedActiveVariations.add(details.getCurrentVariantId());
                        }

                        TuneEventBus.post(new TuneSessionVariableToSet(TUNE_ACTIVE_VARIATION_ID, details.getCurrentVariantId(), TuneSessionVariableToSet.SaveTo.PROFILE));
                        break;
                    }
                }
            } else if (type != null && type.equals(TuneExperimentDetails.DETAIL_TYPE_IN_APP)) {
                TuneInAppMessageExperimentDetails details = new TuneInAppMessageExperimentDetails(experiment);
                inAppMessageExperimentDetailsTemp.put(TuneJsonUtils.getString(experiment, TuneExperimentDetails.DETAIL_EXPERIMENT_NAME_KEY), details);
            }
        }

        setPhookExperimentDetails(new HashMap<String, TunePowerHookExperimentDetails>(phookExperimentDetailsTemp));
        setInAppExperimentDetails(new HashMap<String, TuneInAppMessageExperimentDetails>(inAppMessageExperimentDetailsTemp));
    }

    private synchronized void setPhookExperimentDetails(Map<String, TunePowerHookExperimentDetails> phookExperimentDetails) {
        this.phookExperimentDetails = phookExperimentDetails;
    }

    private synchronized void setInAppExperimentDetails(Map<String, TuneInAppMessageExperimentDetails> inAppExperimentDetails) {
        this.inAppExperimentDetails = inAppExperimentDetails;
    }

    public synchronized Map<String, TunePowerHookExperimentDetails> getPhookExperimentDetails() {
        return new HashMap<String, TunePowerHookExperimentDetails>(phookExperimentDetails);
    }

    public synchronized Map<String, TuneInAppMessageExperimentDetails> getInAppExperimentDetails() {
        return new HashMap<String, TuneInAppMessageExperimentDetails>(inAppExperimentDetails);
    }

}

File: sdk-ios/Tune/Tune/TuneExperimentManager.m
//
//  TuneExperimentManager.m
//  TuneMarketingConsoleSDK
//
//  Created by Charles Gilliam on 9/29/15.
//  Copyright © 2015 Tune. All rights reserved.
//

#import "TuneExperimentManager.h"
#import "TuneSkyhookCenter.h"
#import "TunePlaylist.h"
#import "TuneExperimentDetails+Internal.h"
#import "TuneInAppMessageExperimentDetails+Internal.h"
#import "TunePowerHookExperimentDetails+Internal.h"
#import "TunePowerHookValue.h"
#import "TuneState.h"
#import "TuneAnalyticsConstants.h"


@implementation TuneExperimentManager

NSDictionary *_powerHookExperimentDetails;
NSDictionary *_inAppExperimentDetails;
NSObject *dictLock;

#pragma mark - Initialization

- (id)initWithTuneManager:(TuneManager *)tuneManager {
    self = [super initWithTuneManager:tuneManager];
    
    if (self) {
        dictLock = [[NSObject alloc] init];
        [self reset];
    }
    
    return self;
}

- (void)bringUp {
    [self registerSkyhooks];
}

- (void)bringDown {
    [self unregisterSkyhooks];
    [self reset];
}

- (void)reset {
    @synchronized(dictLock) {
        _powerHookExperimentDetails = [[NSDictionary alloc] init];
        _inAppExperimentDetails = [[NSDictionary alloc] init];
    }
}

#pragma mark - Skyhook Registration

- (void)registerSkyhooks {
    [self unregisterSkyhooks];
    
    // Updating the experiment details should happen before the power hooks so that the current experiments can be checked in the power hook callback
    [[TuneSkyhookCenter defaultCenter] addObserver:self
                                          selector:@selector(handlePlaylistChanged:)
                                              name:TunePlaylistManagerCurrentPlaylistChanged
                                            object:nil
                                          priority:TuneSkyhookPrioritySecond];
}

#pragma mark - Skyhook Handlers

- (void)handlePlaylistChanged:(TuneSkyhookPayload *)payload {
    TunePlaylist *activePlaylist = [payload userInfo][TunePayloadNewPlaylist];
    
    NSMutableDictionary *powerHookExperimentDetailsTemp = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *inAppMessageExperimentDetailsTemp = [[NSMutableDictionary alloc] init];
    
    NSDictionary *experimentDetails = activePlaylist.experimentDetails;
    [experimentDetails enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
        NSString *experimentId = key;
        NSDictionary *experiment = obj;
        
        // Find the appropriate power hook to go with the detail block
        // TODO: we may want to run this check against the active experiments.
        NSString *type = experiment[DetailDictionaryExperimentTypeKey];
        
        if ([type isEqualToString:DetailDictionaryTypePowerHook]) {
            NSDictionary *hooks = activePlaylist.powerHooks;
            for (NSString *hookId in hooks) {
                // Create the PowerHookValue temporarily so we can get some information from it.
                TunePowerHookValue *hookValue = [[TunePowerHookValue alloc] initWithDictionary:hooks[hookId]];
                if ([hookValue.experimentId isEqualToString:experimentId]) {
                    TunePowerHookExperimentDetails *details = [[TunePowerHookExperimentDetails alloc] initWithDetailsDictionary:experiment andPowerHookValue:hookValue];
                    powerHookExperimentDetailsTemp[hookId] = details;
                    [[TuneSkyhookCenter defaultCenter] postSkyhook:TuneSessionVariableToSet
                                                            object:nil
                                                          userInfo:@{ TunePayloadSessionVariableName:TUNE_ACTIVE_VARIATION_ID,
                                                                      TunePayloadSessionVariableValue:[details currentVariantId],
                                                                      TunePayloadSessionVariableSaveType:TunePayloadSessionVariableSaveTypeProfile }];
                    break;
                }
            }
        } else if ([type isEqualToString:DetailDictionaryTypeInApp]) {
            TuneInAppMessageExperimentDetails *details = [[TuneInAppMessageExperimentDetails alloc] initWithDetailsDictionary:experiment];
            inAppMessageExperimentDetailsTemp[experiment[DetailDictionaryExperimentNameKey]] = details;
        }
    }];

    @synchronized(dictLock) {
        _powerHookExperimentDetails = [powerHookExperimentDetailsTemp copy];
        _inAppExperimentDetails = [inAppMessageExperimentDetailsTemp copy];
    }
}

# pragma mark - Experiment Details

- (NSDictionary *)getPowerHookVariableExperimentDetails {
    NSDictionary *result = nil;
    @synchronized(dictLock) {
        result = [_powerHookExperimentDetails copy];
    }
    return result;
}

- (NSDictionary *)getInAppMessageExperimentDetails {
    NSDictionary *result = nil;
    @synchronized(dictLock) {
        result = [_inAppExperimentDetails copy];
    }
    return result;
}

@end
